README.md
# PBot - An AdventureLand bot.
###### DV: @version 0.27.0_028_85a83f5_2020-03-08_20:02:42
## Author: [Joseph Pahl]
## Version: @tag 0.29.0

[Changelog](CHANGELOG.md)
    
## Repository: https://github.com/phanku/PBot

## Description
 
[AdventureLand] is a MMORPG game in which you write JavaScript, also known as `CODE`, that controls your character(s).

[PBot] is my personal [AdventureLand] bot that is currently being developed for playing the game. 

I personally find [AdventureLand] fun as I grew up playing Multi User Dungeons (MUD) back in the day. MUDs are a text 
base role playing game. Scripting for [AdventureLand] reminds me of those days of playing on 
a MUD called [Three Kingdoms].

AdventureLand website: [https://adventure.land/](https://adventure.land/)

------------------------------------------------------------
IDE
------------------------------------------------------------

I am a professional developer with over 10 years of experience in creating web based applications. As such I do love
and have a tendency to use professional IDEs. I used PHPStorm to create this project and have no intentions to ever 
stop. With that said, I also do not plan to assist anyone in using this bot if I ever make it public. PBot comes
as it is and I do not promise anything other than here is the bot that is being developed to play the AdventureLand game.  

I use the file watchers within PHPStorm to automatically execute webpack to compile the various files together into
one file. 

------------------------------------------------------------
How to get started
------------------------------------------------------------

As stated below, this project uses [Webpack] to compile the project into one file for use within [AdventureLand].
It is highly recommended to use a professional IDE while contributing/working with this project. 

Requirements:
    - NodeJS must be installed and functioning on the machine used to compile this project.
    - A Node based package manager must be installed and working on the machine to compile this project. 

1) Clone the this repository on to a local machine. 
2) Open a terminal to the root of the project. 
3) Type `npm install` or `yarn install` to install the needed development dependencies.
4) Execute webpack. 
5) Copy the generated `Pbot.js` file located in the `dist` folder into your clipboard.
6) Paste the code into your `CODE` section of AdventureLand.

------------------------------------------------------------
Auto Version for files
------------------------------------------------------------

The version, and since, identifiers within the file documentation blocks as well as the verison identifier for this
readme are all done automatically using a hook within the [Git] system. 

The versions of files within this application was done by the `pre-commit` script that is located
in the `.githooks` folder. 

Note: The GIT based auto version system only works on the following file types: 

    - PHP   :   *.php
    - CSS   :   *.css
    - SCSS  :   *.scss
    - JS    :   *.js
    - MD    :   *.md
    - HBS   :   *.hbs

To activate the auto versioning:
 1) Clone this repo.
 2) Type the following command in the command line while being within the local repository folder:
 
 `git _config core.hooksPath ./githooks`
  
 NOTE: You will have to execute the `git commit` commands via the terminal using [PhpStorm] due to an un-known bug
 in the `pre-commit` file for `git`. All other version control system (VCS) options within [PhpStorm] will still work
 as intended. 

The master branch of the repository should only hold working versions of PBot. 

Instructions for pushing a new version to repository: 
1) Make the changes to the scripts as needed. 
2) Type `git tag -a -m "<Version note>" <version number>`
3) Type `git add .`
4) Type `git commit -m "<commit message>"`
5) Type `git push --follow-tags`
 
------------------------------------------------------------
Webpack 
------------------------------------------------------------ 
    
The [Yarn Package Manager] (Yarn) is used to manage the JavaScript packages, that are contained within the [NPM] repo, 
used within PBot. It is recommended that [Yarn Package Manager] is used however there is a possibility that [NPM] 
could also be used.
 
The front end packages are found under the `node_modules` directory but the actual `package.json` file 
is located in the root folder. 

- Requirements: 
    - [NodeJs]
    - [NPM]
    - [Yarn Package Manager]
    
This section describes on how to add, remove, or update the packages within the `node_modules` folder.
All of the instructions below are executed on a terminal/command line. 

- Update: 
    
    - In the root folder type: 
        - NPM: `npm update`,
        - Yarn: `yarn install`
    
- Add/install:

    - In the root folder type: 
        - NPM: `npm install <package name>`
        - Yarn: `yarn add <package name> --dev` or `yarn add <package name>`
    
- Remove/uninstall

    - In the root folder type: 
        - NPM: `npm uninstall <package name`
        - Yarn: `yarn add <package name> --dev` or `yarn add <package name>`
 
------------------------------------------------------------
##### PhpStorm File watchers
------------------------------------------------------------

It is highly recommended that to use [PhpStorm]'s file watchers while developing PBot.

Three file watchers will need to be set up

- SCSS : Watches for changes to [SASS] files.
- JS : Watches for changes to `JavaScript` files.
- HBS : Watches for changes to `hbs` files.
    
The file watcher settings below are for setting up [Webpack] to automatically fire `yarn run build` when a file type
listed above changes. Using [PhpStorm]'s file watcher, [Webpack], and [PhpStorm]'s auto deployment combined delivers a
decent development environment. 

1. Open the settings dialog in [PhpStorm]. 
2. Find and unfold the `Tools` settings in the left side of the dialog window.
3. Click on `File Watchers` settings in the left side of the dialog window.
4. Three `File Watchers` need to be set up, one each for `js`, `hbs`, and `scss`.
    1. Click on the `+` button on the right side of the dialog window. This will create a new `File watcher`.
    2. Select `<custom>` option from the context menu that displays. 
    3. Name the `File Watcher` so that represents one of the three types above.
    4. Select either `JavaScript`, `SCSS Style Sheet`, or `HTML` depending on which `File Watcher` you are establishing.
    5. Select the `/src/` folder and then click the `Include Recursively` button. The `Pattern` field should appear.
    6. Click `OK`.
    7. Set the `Program` in the `Edit Watcher` dialog window to `yarn`.
    8. Set the `Arguments` to `run build` in the `Edit Watcher` dialog window. 
    9. Set the `Output paths to refresh` to `$ProjectFileDir$`. 
        1. Note: `$ProjectFileDir$` is a [PhpStorm] macro.
    10. Set the `Working Directory` to `$ProjectFileDir$\assets`. 
    11. Click `Ok`.
5. Repeat the steps above once for `js`, `hbs`, and `scss`. 

[//]: # (These are reference links used in the body of this note.)
     
   [Joseph Pahl]: <https://github.com/phanku/>   
   [NodeJs]: <https://nodejs.org/en/>
   [Node Package Manager]: <https://www.npmjs.com/>
   [NPM]: <https://www.npmjs.com/>
   [Yarn Package Manager]: <https://yarnpkg.com/en/>
   [PhpStorm]: <https://www.jetbrains.com/phpstorm/>
   [SASS]: <https://sass-lang.com/>
   [Webpack]: <https://webpack.js.org/>
   [Underscore]: <https://underscorejs.org/>   
   [Git]: <https://git-scm.com/>  
   [PBot]: <https://github.com/phanku/PBot>
   [Three Kingdoms]: <http://3k.org/>
   [AdventureLand]: <https://store.steampowered.com/app/777150/Adventure_Land__The_Code_MMORPG/>


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.29.0](https://github.com/phanku/PBot/compare/0.27.0...0.29.0)

> 10 March 2020

- - Begin adding directive queue. [`b319636`](https://github.com/phanku/PBot/commit/b3196363b477e7a5029cdad763020302c30d99fd)
- Refactored imports. [`a71f985`](https://github.com/phanku/PBot/commit/a71f9851ab328ab9efe9c083713e20d33ccf027a)
- Added directive `UseManaBurst`. [`6328c46`](https://github.com/phanku/PBot/commit/6328c460882cd944dd0098cc79d220d0d0b3bdd4)

#### [0.27.0](https://github.com/phanku/PBot/compare/0.26.0...0.27.0)

> 7 March 2020

- Flushout auto party [`#5`](https://github.com/phanku/PBot/pull/5)
- Added party CM signature and payload feature. [`7a6e8e1`](https://github.com/phanku/PBot/commit/7a6e8e1d613307d638b884457deb12d438fabe76)
- Possible proof of concept. [`c40dbc5`](https://github.com/phanku/PBot/commit/c40dbc5b84796a3b7ec4eb63d0634a11deb8cd2f)
- Pushed. [`13cd16b`](https://github.com/phanku/PBot/commit/13cd16b1eca1da1bf53907c1e1e770c9fd7fd025)

#### [0.26.0](https://github.com/phanku/PBot/compare/0.24.0...0.26.0)

> 26 February 2020

- More accurate dps [`#4`](https://github.com/phanku/PBot/pull/4)
- Updated documentation. [`#3`](https://github.com/phanku/PBot/pull/3)
- Added Stubbed UseSkill directive. [`75f2a70`](https://github.com/phanku/PBot/commit/75f2a7021dbd9a973a3ff95cee12b34aa9e6b838)
- Added Stubbed UseSkill directive. [`aade65b`](https://github.com/phanku/PBot/commit/aade65b9e71362c4c837230604c2cc524a4cc350)
- Refactored potion consumption logic to use game function is_on_cooldown. [`e61296a`](https://github.com/phanku/PBot/commit/e61296a234e4edca018ba4c5421fb8fe4cb1fd82)

#### [0.24.0](https://github.com/phanku/PBot/compare/0.23.0...0.24.0)

> 26 February 2020

- Refactored DPS and GPS to be more concurrent. [`#2`](https://github.com/phanku/PBot/pull/2)
- Update party class module [`#1`](https://github.com/phanku/PBot/pull/1)
- Refactored DPS and GPS to be more concurrant. [`44f84ac`](https://github.com/phanku/PBot/commit/44f84ac0689d212c8de0751ad4d2436100162991)
- Fixed documentation. Added stubbed priest class. [`8a03eec`](https://github.com/phanku/PBot/commit/8a03eec92e2d7dcd0c23cca5b3ef12b2bd3b184e)
- Update README.md [`d572d72`](https://github.com/phanku/PBot/commit/d572d721fbc1b05e96db820aa36f97eb03ceb0bb)

#### [0.23.0](https://github.com/phanku/PBot/compare/0.22.4...0.23.0)

> 24 February 2020

- Rough layout of the party module. [`151adb0`](https://github.com/phanku/PBot/commit/151adb096ebab4fa7d87e92121e9e648366933aa)
- Updated documentation. [`caded2a`](https://github.com/phanku/PBot/commit/caded2a907dc3759089ff2d21d84d3c8abb34347)
- Updated documentation. [`741c2d6`](https://github.com/phanku/PBot/commit/741c2d6a483008369d0f3825e577503a07f21b5d)

#### [0.22.4](https://github.com/phanku/PBot/compare/0.22.3...0.22.4)

> 24 February 2020

- Added reference to `3k.org` for `Three Kingdoms` text in `README.md`. [`8f1313f`](https://github.com/phanku/PBot/commit/8f1313f3ba1ea539aeff67d752fa3e9494fb244d)

#### [0.22.3](https://github.com/phanku/PBot/compare/0.22.2...0.22.3)

> 24 February 2020

- Removed commented code. [`b2ab3f2`](https://github.com/phanku/PBot/commit/b2ab3f2a9f8c36243bf562c711c87eba4a446d91)

#### [0.22.2](https://github.com/phanku/PBot/compare/0.22.0...0.22.2)

> 24 February 2020

- Initial commit from private repository. [`8d4ae08`](https://github.com/phanku/PBot/commit/8d4ae081201de98dabf506e88fc0d23e89f57e30)
- Fixed link to changelog in readme. [`19887de`](https://github.com/phanku/PBot/commit/19887de02231412f5a0dfcff956628227a34cf4d)

#### 0.22.0

> 24 February 2020

- Initial commit [`d08612d`](https://github.com/phanku/PBot/commit/d08612d10dc841be167d8f000224aba35a7ea5db)


src/game/hud/Hud.js
/**
 * Provides the PBot HUD.
 * @author Joseph Pahl <https://github.com/phanku/>
 * @version 0.27.0_032_c75b903_2020-03-10_16:26:52
 * @since 0.22.0_001_d08612d_2020-02-24_08:53:57
 * @todo Move any settings that could be considered semi-permanent into the storage module/local storage.
 */

// Imports
import { C } from "../Constants";
import Logger from "../logger/Logger";
import Storage from "../Storage";
import Communications from "../Communications";

/**
 * Provides the methods and HUD interface.
 */
export default class Hud {

    /**
     * Loads the templates into this object.
     */
    static loadTemplates() {
        this._templates = {
            main: require('./main/main.hbs')
        };

        Logger.log(Logger.DEBUG, 'Hud Templates loaded.');
    }

    /**
     * Determines if the HUD has been added to the DOM.
     * @return {boolean}
     */
    static isHudPresent() {
        let present = parent.$('#pbot-hud').length > 0;
        Logger.log(Logger.DEBUG, 'Hud Present: ' + (present));
        return present;
    }

    /**
     * Builds and injects the hud into the DOM.
     */
    static buildHud() {
        parent.$('body').append(this._templates.main());
        parent.$('head').append($('<style></style>').text(require('./main/main.scss')));
        Logger.log(Logger.DEBUG, 'Hud built.');
    }

    /**
     * Fetches and stores references to the HUD's elements.
     */
    static fetchHudReferences() {
        this._R.hud = parent.$('#pbot-hud');
        this._R.hudTabs = parent.$('#pbot-hud-tabs');
        this._R.hudTabButtons = parent.$('#pbot-hud-tabs .tab-button');
        this._R.hudTabContents = parent.$('#pbot-hud-tabs .tab-content');
        this._R.autoAttack = parent.$('#auto-attack');
        this._R.closestTarget = parent.$('#closest-target');
        this._R.potionCountHP = parent.$('#hp-number');
        this._R.potionCountMP = parent.$('#mp-number');
        this._R.damagePerSecond = parent.$('#dps-number');
        this._R.XPPerSecond = parent.$('#xpps-number');
        this._R.goldPerSecond = parent.$('#gps-number');
        this._R.gold = parent.$('#gold-number');
        this._R.eta2l = parent.$('#eta2l-time');
        this._R.maxAtt = parent.$('#max-att');
        this._R.hpPotionTrigger = parent.$('#hp-potion-trigger');
        this._R.mpPotionTrigger = parent.$('#mp-potion-trigger');
        this._R.hudMinimizeButton = parent.$('#pbot-minimize');
        this._R.hudContentContainer = parent.$('#pbot-hud-content');
        this._R.hudGraphDps = parent.document.getElementById('pbot-graph-dps');
        this._R.hudGraphGps = parent.document.getElementById('pbot-graph-gps');
        this._R.hudGraphXPps = parent.document.getElementById('pbot-graph-xpps');
        Logger.log(Logger.DEBUG, 'Hud elements references stored.');
        // this.setEventListeners();
    }

    /**
     * Sets the bot version in the hud title.
     * @param version
     */
    static setBotVersion(version) {
        parent.$('#pbot-bot-version').text(version);
        Logger.log(Logger.DEBUG, 'Bot version added to hud.');
    }

    /**
     * Returns an integer based on if stat1 is larger than, less than, or equal to stat2.
     * @param stat1 The first stat.
     * @param stat2 The second stat.
     * @returns {number}
     */
    static compare(stat1, stat2) {
        return (stat1 > stat2 ? 1 : stat1 < stat2 ? -1 : 0);
    }

    /**
     * Causes the HUD to be draggable by using the jQuery UI already installed in the game.
     */
    static initHudDragging() {
        let hudLocation = this.getHudSetting('location');
        this._R.hud.css('left', hudLocation.left).css('top', hudLocation.top);
        this._R.hud.draggable({
            stop: (e, ui) => {
                Hud.setHudSetting('location', {
                    top: ui.position.top,
                    left: ui.position.left
                })
            },
        });
        Logger.log(Logger.DEBUG, 'Hud dragging initialized.');
    }

    /**
     * Activates the behavior for the tabs within the HUD.
     */
    static initHudTabs() {
        // Loading the HUD tab settings from local storage.
        let hudActiveTab = this.getHudSetting('activeTab');
        // Setting the current active tabs according to the storage module.
        this._R.hudTabButtons.find('[data-tab="' + hudActiveTab + '"]').addClass('active');
        this._R.hudTabButtons.find('#' + hudActiveTab).addClass('active');

        // Activating the event bindings.
        this._R.hudTabButtons.on('click', (e) => {

            // Saving the tab choice to local storage.
            this.setHudSetting('activeTab', window.parent.$(e.currentTarget).data('tab'));

            this._R.hudTabContents.each((i,v) => {
                window.parent.$(v).removeClass('active');
            });

            this._R.hudTabButtons.removeClass('active');
            window.parent.$(e.currentTarget).addClass('active');
            window.parent.$('#pbot-hud-tabs #' + window.parent.$(e.currentTarget).data('tab')).addClass('active');
        });

        Logger.log(Logger.DEBUG, 'Hud tabs initialized.');
    }

    /**
     * Activates the Pretty check boxes in the hud.
     */
    static initHudChecks() {
        this._R.autoAttack.on('change', (e) => {
            Communications.broadcast(
                Communications.getBroadcastObject(C.COMMS.HUD.SETTING_UPDATE, this.getCharacterConfigs(), this));
        });

        Logger.log(Logger.DEBUG, 'Hud pretty checks initialized.');
    }

    /**
     * Activates the blur event in the hud that will cause the potion consumption points to be updated.
     */
    static initTriggerPoints() {
        this._R.mpPotionTrigger.add(this._R.hpPotionTrigger).on('blur', (e) => {
            Communications.broadcast(
                Communications.getBroadcastObject(C.COMMS.HUD.SETTING_UPDATE, this.getCharacterConfigs(), this));
        });
    }

    /**
     * Returns an object that contains the settings in the HUD.
     */
    static getCharacterConfigs() {
        // Collect and return the Hud configurations.
        return {
            attackActive: this._R.autoAttack.prop('checked'),
            closestTarget: this._R.closestTarget.prop('checked'),
            maxAtt: this._R.maxAtt.val(),
            hpPotionTrigger: this._R.hpPotionTrigger.val(),
            mpPotionTrigger: this._R.mpPotionTrigger.val(),
        };
    }

    /**
     * Stores reference to the character object in this class.
     * @param character
     */
    static setCharacter(character) {
        this._character = character;
        Logger.log(Logger.DEBUG, 'Character reference stored.');
    }

    /**
     * Saves the HUD setting to the local storage.
     * @param key
     * @param value
     */
    static setHudSetting(key, value) {
        Storage.set('hud_' + key, value);
    }

    /**
     * Fetches the HUD setting from the local storage.
     * @param key
     * @param value
     * @return {*}
     */
    static getHudSetting(key, value) {
        return Storage.get('hud_' + key);
    }

    /**
     * Initializes the HUD settings in local storage.
     */
    static initHudSettings() {
        if (this.getHudSetting('init')) {
            return;
        }

        this.setHudSetting('activeTab', 'pbot-hud-stats');
        this.setHudSetting('location', {top: 160, left: 0});
        this.setHudSetting('minimized', false);
        this.setHudSetting('init', true);
        Logger.log(Logger.DEBUG, "Hud settings initialized.");
    }

    /**
     * Initializes the HUD minimization feature.
     */
    static initHudMinimize() {
        this._R.hudMinimizeButton.on('click', (e)=>{
            if (this._R.hudContentContainer.hasClass('folded')) {
                this.setHudSetting('minimized', false);
                this._R.hudContentContainer.removeClass('folded');
            } else {
                this.setHudSetting('minimized', true);
                this._R.hudContentContainer.addClass('folded');
            }
        });

        // Loading the last saved HUD minimized settings.
        if (this.getHudSetting('minimized')) {
            this._R.hudContentContainer.addClass('folded');
        }

        Logger.log(Logger.DEBUG, "Hud minimization initialized.");
    }

    /**
     * Updates the potions section of the HUD stats.
     * @param potions An object containing the character's current potion levels.
     */
    static updatePotions(potions) {
        this._R.potionCountHP.text(potions.HP);
        this._R.potionCountMP.text(potions.MP);
    }

    /**
     * Updates the DPS on the Hud.
     * @param stats An object with the stats.
     */
    static updateDPSStat(stats) {
        this._R.damagePerSecond.html(stats.dps);
        this._top.dps = stats.dps > this._top.dps ? stats.dps : this._top.dps;
        this._graphDps.data.datasets[0].data = [this._top.dps];
        this._graphDps.data.datasets[1].data = [stats.dps];
        this._graphDps.update();
    }

    /**
     * Causes the HUD to update the per second stats.
     * @param stats An object containing the stats.
     */
    static updateHudStats(stats) {
        this.updateDPSStat(stats);
        this._R.XPPerSecond.html(stats.xpps);
        this._R.goldPerSecond.html(stats.gps);
        this._R.gold.html(stats.gold);
        this._R.eta2l.html(stats.eta2l);
        this._top.gps = stats.gps > this._top.gps ? stats.gps : this._top.gps;
        this._top.xpps = stats.xpps > this._top.xpps ? stats.xpps : this._top.xpps;

        this._graphGps.data.datasets[0].data = [this._top.gps];
        this._graphGps.data.datasets[1].data = [stats.gps];
        this._graphGps.update();

        this._graphXPps.data.datasets[0].data = [this._top.xpps];
        this._graphXPps.data.datasets[1].data = [stats.xpps];
        this._graphXPps.update();

        this._last = stats;
    }

    /**
     * Handles the payload of a broad cast message.
     * @param broadcast The broadcast message.
     */
    static broadcastHandler(broadcast) {
        switch(broadcast.context) {
            case C.COMMS.PARTY.INITIALIZED:
                // The party module has acknowledged the HUD being ready.
                // The party module should have sent the current party leader and members here.
                break;

            case C.COMMS.STATS.DAMAGE_PER_SECOND:
                // The stats module has issued an updated DPS number.
                this.updateDPSStat(broadcast.payload);
                break;

            case C.COMMS.CHARACTER.POTION_CONSUMPTION:
                // The character class has issued an character consuming a potion broadcast.
                this.updatePotions(broadcast.payload);
                break;

            case C.COMMS.STATS.UPDATE:
                // The stats module has issued a stats numbers updated broadcast.
                this.updateHudStats(broadcast.payload);
                break;

            case C.COMMS.GAME.BOT_VERSION:
                // PBot version.
                this.setBotVersion(broadcast.payload.version);
                break;
        }
    }

    /**
     * Initializes the canvas graphs.
     * @todo Rename this method to follow naming conventions.
     */
    static startGraph() {
        // The options for the graphics in the HUD.
        let graphOptions = {
            responsive: true,
            legend: {
                display: false
            },
            scales: {
                // xAxes: [{
                //     categoryPercentage: 1,
                //     barPercentage: 0.4
                // }],
                yAxes: [{
                    gridLines: {
                        color: 'rgba(255, 255, 255, 0.4)'
                    },
                    ticks: {
                        beginAtZero: true
                    }
                }]
            }
        };

        this._graphDps = new Chart(this._R.hudGraphDps, {
            type: 'bar',
            data: {
                labels: ['DPS'],
                categoryPercentage: 1,
                datasets: [
                    {
                        data: [0],
                        backgroundColor: 'rgba(255,204,204,1)',
                        barPercentage: 0.4
                    },
                    {
                        data: [0],
                        backgroundColor: 'rgba(204,204,255,1)',
                        barPercentage: 0.4
                    }
                ],
            },
            options: graphOptions
        });

        this._graphGps = new Chart(this._R.hudGraphGps, {
            type: 'bar',
            data: {
                labels: ['GPS'],
                categoryPercentage: 1,
                datasets: [
                    {
                        data: [0],
                        backgroundColor: 'rgba(255,204,204,1)',
                        barPercentage: 0.4
                    },
                    {
                        data: [0],
                        backgroundColor: 'rgba(204,204,255,1)',
                        barPercentage: 0.4
                    }
                ]
            },
            options: graphOptions
        });

        this._graphXPps = new Chart(this._R.hudGraphXPps, {
            type: 'bar',
            data: {
                labels: ['XPPS'],
                categoryPercentage: 1,
                datasets: [
                    {
                        data: [0],
                        backgroundColor: 'rgba(255,204,204,1)',
                        barPercentage: 0.4
                    },
                    {
                        data: [0],
                        backgroundColor: 'rgba(204,204,255,1)',
                        barPercentage: 0.4
                    }
                ]
            },
            options: graphOptions
        });
    }

    /**
     * Initializes the hud.
     */
    static init() {
        // Initialize the references object.
        this._R = {};
        // Store the last stats so the HUD can show differences.
        this._last = {};
        // Store the last
        this._top = {
            dps: 0,
            gps: 0,
            xpps: 0,
        };

        // Load the HUD templates.
        this.loadTemplates();
        // Initialize the HUD settings in the local storage.
        this.initHudSettings();

        // Subscribing to the  object.
        Communications.subscribe((broadcast) => {
            // The broadcastHandler function must be wrapped by a lambda function so as to maintain the
            // reference to this object.
            this.broadcastHandler(broadcast);
        });

        // Checking to see if the HUD is already in the DOM.
        if (!this.isHudPresent()) {
            // The HUD is not in the DOM so now build and inject.
            parent.$.when(
                this.buildHud()
            ).then(()=>{
                parent.$.when(
                    // Fetch references to the sections in the HUD that need to be updated.
                    this.fetchHudReferences()
                ).then(()=>{
                    // Make the hud draggable.
                    this.initHudDragging();
                    // Activate the HUD tabs.
                    this.initHudTabs();
                    this.initHudChecks();
                    this.initTriggerPoints();
                    this.initHudMinimize();
                    setTimeout(()=>{
                        this.startGraph();

                        // Notify any modules waiting for the HUD to come online.
                        Communications.broadcast(
                            Communications.getBroadcastObject(C.COMMS.HUD.INITIALIZED, {}, this));

                    }, 500);
                });

                // this.relocateHud();
            });
        } else {
            // The HUD already exists in the DOM.
            // Fetch references to the sections in the HUD that need to be updated.
            this.fetchHudReferences();
        }

        Logger.log(Logger.DEBUG, 'HUD initialized.');
    }
};

Hud.init();